### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by danillopis**

### CODE REVIEW

File: Assignment4/d.llopis-21C033/task07.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+q6 = prepareQuery('''
+    SELECT ?entityName
+    WHERE {
+        ?entity rdf:type ?type .
+        ?entity ns:knows ?person1 .
+        ?entity ns:knows ?person2 .
+        FILTER (?person1 != ?person2) .
+        ?entity ns:name ?entityName .
+    }''',
```
Comment (SerPablo) [inline Assignment4/d.llopis-21C033/task07.py:135]: This query does not retrieve anything

**Review by SerPablo — COMMENTED** [review]
Please review results for tasks 7.4 and 7.5. Im not able to reproduce them

**Review by danillopis — COMMENTED** [review]
creo que ya está corregido

**Comment by dgarijo** [comment]
- @danillopis Remove the .DSStore and .Rhistory files from the repo.
- Have you run the query? Because you are defining the foaf vocabulary in SPARQL, but you are not declaring the namespace anywhere. The query should not run. Run the query and then review the PR

**Comment by danillopis** [comment]
I don't know how to remove those files just cause they don't appear on my own repository. And yes, I ran the code before on gcollab and it apparently run.

**Comment by danillopis** [comment]
Maybe those files were created after the changes, i don't know why

**Comment by dgarijo** [comment]
Those files are generated in your machine, otherwise they would not be part of the PR. They are invisible though, so you have to configure your machine to view them

**Comment by dgarijo** [comment]
You have not deleted them from your folder but you still have them in the parent one. Also, you are modifying a readme file that you should not

**Comment by dgarijo** [comment]
you are also missing the exercises in RDFlib (you only provide the sparql solution)

**Comment by danillopis** [comment]
I think now is all right. When you can if you can check it, it would be grateful.

File: Assignment4/d.llopis-21C033/task06corr.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+ns3 = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+email_address = Literal("janesmith@gmail.com")
+g.add((JaneSmith, ns3.EMAIL, email_address))
+
+full_name = Literal("Jane Smith")
+g.add((JaneSmith, ns3.FN, full_name))
+
+given_name = Literal("Jane")
+g.add((JaneSmith, ns3.gIVEN, given_name))
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/task06corr.py:72]: GIVEN

File: Assignment4/d.llopis-21C033/task06corr.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+for s, p, o in g:
+  print(s,p,o)
+
+"""**TASK 6.5: Add UPM as the university where John Smith works**"""
+
+# TO DO
+# Visualize the results
+
+ns7 = Namespace("http://somewhere")
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/task06corr.py:85]: why do you keep re-declaring the namespace? Each NS should be declared only once

File: Assignment4/d.llopis-21C033/task06corr.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+"""**Task 6.6: Add that Jown knows Jane using the FOAF vocabulary**"""
+
+# TO DO
+# Visualize the results
+
+foaf = Namespace("http://xmlns.com/foaf/0.1/")
+
+JohnSmith = ns.JohnSmith
+JaneSmith = ns.JaneSmith
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/task06corr.py:100]: Same thing for John and Jane, you already defined them above

File: Assignment4/d.llopis-21C033/trask07corr.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+"""7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL"""
+
+# TO DO
+from rdflib.plugins.sparql import prepareQuery
+
+livingThing_query = prepareQuery('''
+SELECT ?subClase
+WHERE {
+    ?subClase rdfs:subClassOf ns:LivingThing
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/trask07corr.py:37]: You are not listing all subclasses of subclasses of living thing (recursively). Use `*`

File: Assignment4/d.llopis-21C033/trask07corr.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+
+# Create a set to store the subclasses
+subclasses = set()
+
+# Iterate through the triples in the graph
+for subject, predicate, object_ in g.triples((None, RDF.type, RDFS.Class)):
+    for sub, _, sup in g.triples((None, RDFS.subClassOf, subject)):
+        if sup == ns.LivingThing:
+            subclasses.add(sub)
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/trask07corr.py:55]: You are not tracking the full subclass hierarchy, as happened before. Hint: make a recursive function

File: Assignment4/d.llopis-21C033/trask07corr.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+# Create a set to store the individuals of "Person" and its subclasses
+individuals = set()
+
+# Create a set of classes to include "Person" and its subclasses
+classes_to_include = {ns.Person}
+
+# Find subclasses of "Person" and add them to the set of classes to include
+for subject, _, object_ in g.triples((None, RDFS.subClassOf, ns.Person)):
+    classes_to_include.add(subject)
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/trask07corr.py:89]: you are not adding all subclasses. For example, if something is a subclass of a subclass of person

File: Assignment4/d.llopis-21C033/trask07corr.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+for individual in individuals:
+    print("Individual:", individual)
+    for s, p, o in g.triples((individual, None, None)):
+        print(f" Property: {p}, Value: {o}")
+
+query_3 = prepareQuery('''
+SELECT ?individual ?property ?value
+WHERE {
+     ?individual rdf:type/rdfs:subClassOf* ?class .
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/trask07corr.py:122]: With this you are listing the individuals of the subclasses of person. You need to remove `*` in this case

Comment (dgarijo) [reply Assignment4/d.llopis-21C033/trask07corr.py:122]: also, use UNION, not filter, as it's very inefficient here.

File: Assignment4/d.llopis-21C033/trask07corr.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+
+for r in g.query(query_3):
+  print(r.individual, r.property, r.value)
+
+""" 7.4: List the name of the persons who know Rocky"""
+
+q4 = """
+    SELECT ?person
+    WHERE {
+        ?person foaf:knows ns:RockySmith .
```
Comment (dgarijo) [inline Assignment4/d.llopis-21C033/trask07corr.py:140]: you are not listing the name of the persons, just their URI

**Review by dgarijo — CHANGES_REQUESTED** [review]
You need to remove changes to the README, the HandsOn and leave ONLY your d.llopis file.
You are close to the solutions, but need a bit more work

**Comment by danillopis** [comment]
Hi, I think now it's everything OK...

**Comment by dgarijo** [comment]
You are removing the README files of the main repository, which is incorrect. Please only add the files you want us to review

**Comment by dgarijo** [comment]
Also, you are not following the instructions shown in class. You have to create a folder with your id, not replace the files in Assignment4. Once you follow the instructions, let us know

**Comment by danillopis** [comment]
Is it correct now? Sorry for all the inconvenience

File: Assignment4/danillopis - 21C033/trask07corr.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+for r in g.query(person_query):
+  print(r)
+
+
+rdf = Namespace("http://www.w3.org/1999/02/22-rdf-syntax-ns#")
+
+print("RDFLib")
+for s,p,o in g.triples((None, RDF.type, ns.Person)):
+  print(s)
+for s,p,o in g.triples((None, RDFS.subClassOf, ns.Person)):
```
Comment (dgarijo) [inline Assignment4/danillopis - 21C033/trask07corr.py:85]: you have to also do a recursive function here, unless you use your previously define recursive function to get the classes

File: Assignment4/danillopis - 21C033/trask07corr.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+      ?entity foaf:knows ?relatedEntity .
+
+}
+GROUP BY ?entity
+HAVING (COUNT(DISTINCT ?relatedEntity) >= 2)
+
+''',
+initNs = { "foaf": FOAF , "ns" : ns, "vcard": VCARD}
+)
+
```
Comment (dgarijo) [inline Assignment4/danillopis - 21C033/trask07corr.py:166]: missing exercise in rdflib

File: Assignment4/danillopis - 21C033/trask07corr.py [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  print(r)
+
+""" 7.4: List the name of the persons who know Rocky"""
+
+from rdflib import FOAF
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+
+
+for s, p, o in g.triples((None, FOAF.knows, ns.RockySmith)):
+  print(s)
```
Comment (dgarijo) [inline Assignment4/danillopis - 21C033/trask07corr.py:134]: This is not correct. You have to return the name, as you do below

**Review by dgarijo — CHANGES_REQUESTED** [review]
You have it mostly ok, very minor issues pending

**Comment by danillopis** [comment]
I hope it's ok. Sorry for all the inconvenience.

**Comment by anaigmo** [comment]
Entregado y corregido

**Pull request merged**